File: src/spirite/base/image_data/mediums/PrismaticMedium.java
package spirite.base.image_data.mediums;

import jdk.nashorn.internal.runtime.options.Option;
import org.jetbrains.annotations.NotNull;
import spirite.base.graphics.DynamicImage;
import spirite.base.graphics.GraphicsContext;
import spirite.base.graphics.GraphicsContext.Composite;
import spirite.base.graphics.IImage;
import spirite.base.graphics.RawImage;
import spirite.base.image_data.ImageWorkspace;
import spirite.base.image_data.ImageWorkspace.BuildingMediumData;
import spirite.base.image_data.mediums.drawer.DefaultImageDrawer;
import spirite.base.image_data.mediums.drawer.IImageDrawer;
import spirite.base.util.MUtil;
import spirite.base.util.linear.MatTrans;
import spirite.base.util.linear.MatTrans.NoninvertableException;
import spirite.base.util.linear.Rect;
import spirite.base.util.linear.Vec2;
import spirite.hybrid.HybridHelper;
import spirite.hybrid.HybridUtil;
import spirite.hybrid.HybridUtil.UnsupportedImageTypeException;
import spirite.hybrid.MDebug;
import spirite.hybrid.MDebug.ErrorType;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/***
 * PrismaticInternalImages are a type of Internal Image that behave similarly to 
 * Dynamic images, but separate each layer by color.
 */
public class PrismaticMedium implements IMedium {
	public static class LImg {
		// NOTE: Setting all of these to public SHOULD be fine so that
		//	PrismaticInternalImages can easily be externally built and loaded in
		//	but then you have to make sure you never pass the internal LImgs
		public int color;
		public DynamicImage img;
		
		public LImg() {}
		public LImg(LImg other, boolean deepcopy) {
			color = other.color;
			img = (deepcopy)?other.img.deepCopy():other.img;
		}
		public LImg(LImg other, boolean deepcopy, boolean copyForSave_IGNORED) {
			color = other.color;
            img = (deepcopy)?other.img.deepCopy():other.img;
			//img = HybridUtil.copyForSaving(other.img);
		}
	}
	private List<LImg> layers = new ArrayList<>();
	private RawImage compositionImg;
	private boolean compIsBuilt = false;
	private Rect compRect = null;
	private final ImageWorkspace context;
	
	public PrismaticMedium( ImageWorkspace context) {
        this.context = context;
    }
	public PrismaticMedium( List<LImg> toImport,ImageWorkspace context) {
        this.context = context;
		for( LImg limg : toImport) {
			layers.add(new LImg(limg, false));
		}
	}

	@Override
	public int getWidth() {
		buildComposition();
		return compRect.width;
	}

	@Override
	public int getHeight() {
		buildComposition();
		return compRect.height;
	}

	@Override
	public int getDynamicX() {
		buildComposition();
		return compRect.x;
	}

	@Override
	public int getDynamicY() {
		buildComposition();
		return compRect.y;
	}
	private void buildComposition() {
		if( compIsBuilt)
			return;
		
		Rect r = new Rect( 0, 0, 0, 0);
		for( LImg img : layers)
			r = r.union(img.img.getXOffset(), img.img.getYOffset(), img.img.getWidth(), img.img.getHeight());
		compRect = r;
		
		if( compositionImg != null)
			compositionImg.flush();
		if( compRect == null || compRect.isEmpty()) {
			compositionImg = HybridHelper.createNillImage();
		}
		else {
			compositionImg = HybridHelper.createImage(r.width, r.height);
			GraphicsContext gc = compositionImg.getGraphics();
			for( LImg img : layers) {
				gc.drawImage(img.img.getBase(), img.img.getXOffset() - r.x, img.img.getYOffset() - r.y);
			}
		}
		
		compIsBuilt = true;
	}

	public void drawBehind( GraphicsContext gc, int color) {
		for( LImg img : layers) {
			gc.drawImage(img.img.getBase(), img.img.getXOffset() - compRect.x, img.img.getYOffset() - compRect.y);
			if( (img.color & 0xFFFFFF) == (color & 0xFFFFFF)) return;
		}
	}
	public void drawFront(GraphicsContext gc, int color) {
		boolean drawing = false;
		for( LImg img : layers) {
			if( drawing) 
				gc.drawImage(img.img.getBase(), img.img.getXOffset() - compRect.x, img.img.getYOffset() - compRect.y);
			else if( (img.color & 0xFFFFFF) == (color & 0xFFFFFF)) 
				drawing = true;
		}
		
	}

	@Override
	public BuiltMediumData build( BuildingMediumData building) {
		return new PrismaticBuiltImageData( building);
	}

	@Override
	public IMedium dupe() {
		PrismaticMedium pii = new PrismaticMedium(context);
		
		pii.layers = new ArrayList<>(this.layers.size());
		for( LImg img : this.layers) {
			pii.layers.add(new LImg(img, true));
		}
		return pii;
	}
	@Override
	public IMedium copyForSaving() {
		PrismaticMedium pii = new PrismaticMedium(context);
		
		pii.layers = new ArrayList<>(this.layers.size());
		for( LImg img : this.layers) {
			pii.layers.add(new LImg(img, true, true));
		}
		return pii;
	}
	
	public void moveLayer(int draggingFromIndex, int draggingToIndex) {
		layers.add(draggingToIndex, layers.remove(draggingFromIndex));
		compIsBuilt = false;
	}
	
	/** Note: the LImg's it returns are copies.  Changing the internal data will
	 * not change the PrismaticImage's data, but drawing to the image WILL effect
	 * the PrismaticImage. */
	public List<LImg> getColorLayers() {
		ArrayList<LImg> list = new ArrayList<LImg>(layers.size());
		
		for( LImg limg : layers) {
			list.add( new LImg(limg, false));
		}
		
		return list;
	}

	boolean flushed = false;
	@Override
	public void flush() {
		if( !flushed) {
			for( LImg img : layers)
				img.img.flush();
			if( compositionImg != null)
				compositionImg.flush();
			flushed = true;
		}
	}
	@Override
	protected void finalize() throws Throwable {
		flush();
	}

	@Override
	public IImage readOnlyAccess() {
		buildComposition();
		return compositionImg;
	}

	@Override
	public InternalImageTypes getType() {
		return InternalImageTypes.PRISMATIC;
	}

	public class PrismaticBuiltImageData extends BuiltMediumData {
	    private final int workingColor;
	    private PrismaticBuiltImageData( BuildingMediumData building) {
	        super(building);
	        workingColor = building.color;
        }

        @NotNull
        @Override
        public MatTrans getDrawTrans() {
	        return new MatTrans();
        }

        @Override
        public int getDrawWidth() {
            return 0;
        }

        @Override
        public int getDrawHeight() {
            return 0;
        }

        @NotNull
        @Override
        public MatTrans getSourceTransform() {
            return getTrans();
        }

        @Override
        public int getSourceWidth() {
            return getWidth();
        }

        @Override
        public int getSourceHeight() {
            return getHeight();
        }

        @Override
        protected void _doOnGC(@NotNull DoerOnGC doer) {
            prepareLImg().img.doOnGC(doer, getTrans());
        }

        @Override
        protected void _doOnRaw(@NotNull DoerOnRaw doer) {
	        prepareLImg().img.doOnRaw(doer, getTrans());
        }

        private LImg prepareLImg() {
            Optional<LImg> optionalLImg = layers.stream().filter((limg) -> limg.color == workingColor).findAny();
            LImg colorLayer;
            if( optionalLImg.isPresent())
                colorLayer = optionalLImg.get();
            else {
                colorLayer = new LImg();
                colorLayer.img = new DynamicImage(context, null, 0, 0);
                colorLayer.color = workingColor;
                layers.add(colorLayer);
            }
            return colorLayer;
        }
	}

	@Override public IImageDrawer getImageDrawer(BuildingMediumData building) {return new DefaultImageDrawer(this, building);}
}

File: src/spirite/base/image_data/mediums/PrismaticMedium.kt
package spirite.base.image_data.mediums

import jdk.nashorn.internal.runtime.options.Option
import spirite.base.graphics.DynamicImage
import spirite.base.graphics.GraphicsContext
import spirite.base.graphics.GraphicsContext.Composite
import spirite.base.graphics.IImage
import spirite.base.graphics.RawImage
import spirite.base.image_data.ImageWorkspace
import spirite.base.image_data.ImageWorkspace.BuildingMediumData
import spirite.base.image_data.mediums.drawer.DefaultImageDrawer
import spirite.base.image_data.mediums.drawer.IImageDrawer
import spirite.base.util.MUtil
import spirite.base.util.linear.MatTrans
import spirite.base.util.linear.MatTrans.NoninvertableException
import spirite.base.util.linear.Rect
import spirite.base.util.linear.Vec2
import spirite.hybrid.HybridHelper
import spirite.hybrid.HybridUtil
import spirite.hybrid.HybridUtil.UnsupportedImageTypeException
import spirite.hybrid.MDebug
import spirite.hybrid.MDebug.ErrorType

import java.util.ArrayList
import java.util.Optional

/***
 * PrismaticInternalImages are a type of Internal Image that behave similarly to
 * Dynamic images, but separate each layer by color.
 */
class PrismaticMedium : IMedium {
    private var layers: MutableList<LImg> = ArrayList()
    private var compositionImg: RawImage? = null
    private var compIsBuilt = false
    private var compRect: Rect? = null
    private val context: ImageWorkspace

    override val width: Int
        get() {
            buildComposition()
            return compRect!!.width
        }

    override val height: Int
        get() {
            buildComposition()
            return compRect!!.height
        }

    override val dynamicX: Int
        get() {
            buildComposition()
            return compRect!!.x
        }

    override val dynamicY: Int
        get() {
            buildComposition()
            return compRect!!.y
        }

    /** Note: the LImg's it returns are copies.  Changing the internal data will
     * not change the PrismaticImage's data, but drawing to the image WILL effect
     * the PrismaticImage.  */
    val colorLayers: List<LImg>
        get() {
            val list = ArrayList<LImg>(layers.size)

            for (limg in layers) {
                list.add(LImg(limg, false))
            }

            return list
        }

    internal var flushed = false

    override val type: IMedium.InternalImageTypes
        get() = IMedium.InternalImageTypes.PRISMATIC

    class LImg(
            val color: Int,
            val img: DynamicImage
    ) {
        constructor(other: LImg, deepcopy: Boolean) : this(
                other.color,
                if( deepcopy) other.img.deepCopy() else other.img
        ){}

        constructor(other: LImg, deepcopy: Boolean, copyForSave_IGNORED: Boolean) :this(
                other.color,
                if( deepcopy) other.img.deepCopy() else other.img
        ){
            //img = HybridUtil.copyForSaving(other.img);
        }
    }

    constructor(context: ImageWorkspace) {
        this.context = context
    }

    constructor(toImport: List<LImg>, context: ImageWorkspace) {
        this.context = context
        for (limg in toImport) {
            layers.add(LImg(limg, false))
        }
    }

    private fun buildComposition() {
        if (compIsBuilt)
            return

        var r = Rect(0, 0, 0, 0)
        for (img in layers)
            r = r.union(img.img.xOffset, img.img.yOffset, img.img.width, img.img.height)
        compRect = r

        if (compositionImg != null)
            compositionImg!!.flush()
        if (compRect == null || compRect!!.isEmpty) {
            compositionImg = HybridHelper.createNillImage()
        } else {
            compositionImg = HybridHelper.createImage(r.width, r.height)
            val gc = compositionImg!!.graphics
            for (img in layers) {
                gc.drawImage(img.img.base, img.img.xOffset - r.x, img.img.yOffset - r.y)
            }
        }

        compIsBuilt = true
    }

    fun drawBehind(gc: GraphicsContext, color: Int) {
        for (img in layers) {
            gc.drawImage(img.img.base, img.img.xOffset - compRect!!.x, img.img.yOffset - compRect!!.y)
            if (img.color and 0xFFFFFF == color and 0xFFFFFF) return
        }
    }

    fun drawFront(gc: GraphicsContext, color: Int) {
        var drawing = false
        for (img in layers) {
            if (drawing)
                gc.drawImage(img.img.base, img.img.xOffset - compRect!!.x, img.img.yOffset - compRect!!.y)
            else if (img.color and 0xFFFFFF == color and 0xFFFFFF)
                drawing = true
        }

    }

    override fun build(building: BuildingMediumData): BuiltMediumData {
        return PrismaticBuiltImageData(building)
    }

    override fun dupe(): IMedium {
        val pii = PrismaticMedium(context)

        pii.layers = ArrayList(this.layers.size)
        for (img in this.layers) {
            pii.layers.add(LImg(img, true))
        }
        return pii
    }

    override fun copyForSaving(): IMedium {
        val pii = PrismaticMedium(context)

        pii.layers = ArrayList(this.layers.size)
        for (img in this.layers) {
            pii.layers.add(LImg(img, true, true))
        }
        return pii
    }

    fun moveLayer(draggingFromIndex: Int, draggingToIndex: Int) {
        layers.add(draggingToIndex, layers.removeAt(draggingFromIndex))
        compIsBuilt = false
    }

    override fun flush() {
        if (!flushed) {
            for (img in layers)
                img.img.flush()
            if (compositionImg != null)
                compositionImg!!.flush()
            flushed = true
        }
    }

    @Throws(Throwable::class)
    protected fun finalize() {
        flush()
    }

    override fun readOnlyAccess(): IImage {
        buildComposition()
        return compositionImg!!
    }

    inner class PrismaticBuiltImageData constructor(building: BuildingMediumData) : BuiltMediumData(building) {
        private val workingColor: Int = building.color

        override val drawTrans: MatTrans get() = MatTrans()
        override val drawWidth: Int get() = context.width
        override val drawHeight: Int get() = context.height

        override val sourceTransform: MatTrans get() = trans
        override val sourceWidth: Int get() = width
        override val sourceHeight: Int get() = height

        override fun _doOnGC(doer: DoerOnGC) {
            prepareLImg().img.doOnGC(doer, trans)
        }

        override fun _doOnRaw(doer: DoerOnRaw) {
            prepareLImg().img.doOnRaw(doer, trans)
        }

        private fun prepareLImg(): LImg {
            val optionalLImg = layers.stream().filter { limg -> limg.color == workingColor }.findAny()
            val colorLayer: LImg
            if (optionalLImg.isPresent)
                colorLayer = optionalLImg.get()
            else {
                colorLayer = LImg(
                        workingColor,
                        DynamicImage(context, null, 0, 0))
                layers.add(colorLayer)
            }
            return colorLayer
        }
    }

    override fun getImageDrawer(building: BuildingMediumData): IImageDrawer {
        return DefaultImageDrawer(this, building)
    }
}
